mmengine/hooks/ema_hook.py

```diff
@@ -175,18 +179,23 @@ def after_load_checkpoint(self, runner, checkpoint: dict) -> None:
             # The original model parameters are actually saved in ema
             # field swap the weights back to resume ema state.
             self._swap_ema_state_dict(checkpoint)
-            self.ema_model.load_state_dict(
-                checkpoint['ema_state_dict'], strict=self.strict_load)
+            _load_checkpoint_to_model(
+                self.ema_model.module,
```

Comment:
The ema_state_dict not only contains model weights but also contains the `steps` and other states in the averaged model (See mmengine/model/averaged_model.py).

So we need to load the ema_state_dict to self.ema_model instead of its module.

Reply:
I understood it and I thought there are three patterns. What do you think?

```
if 'ema_state_dict' in checkpoint and self.resume:
    # Cases in which training is resumed in the middle of the step.
    # In this part, step, ema_state_dict and other params should be loaded.
    self.ema_model.load_state_dict(
                checkpoint['ema_state_dict'], strict=self.strict_load)
elif 'ema_state_dict' in checkpoint:
    # Cases in which train other dataset with the pretrained weight. The pretrain weight has ema_state_dict, but training should start from step 0.
    # In this part, only ema_state_dict should be loaded.
    _load_checkpoint_to_model(
                self.ema_model.module,
                checkpoint['ema_state_dict'],
                False,
                revise_keys=revise_keys)
else:
    # Cases in which the pretrained weight doesn't have ema_state_dict.
    _load_checkpoint_to_model(
                self.ema_model.module,
                copy.deepcopy(checkpoint['state_dict']),
                False,
                revise_keys=revise_keys)
```

Reply:
The second case appears to be rarer. 
Usually, when we need to load a model trained with EMA as pre-trained weights, we will directly load the weights of EMA into the original model and start EMA again. 
In this case, we do not need to load the EMA weights into the EMA model, because **when `step=0`, `AveragedModel` will copy the original model parameters**.

So there are only two cases: Resume or not resume.
If resume, the ema_state_dict should be loaded.
If not resume, directly load the `state_dict`.